### ISSUE TITLE
Subtitle string filtering

### ISSUE THREAD

**Issue by grishka** [iina/iina#3958]
**What you want IINA to do:**

Some downloads come with "SDH" subtitles that include textual descriptions of sounds that are not speech. These descriptions are annoying, and there is no way to get rid of them. However, they *usually* are formatted differently than speech, for example in brackets like `[ distant steps approaching ]` or `[ theme music playing ]`. It would be nice to have a setting where I can specify e.g. a regular expression that subtitle strings would be tested against, and not displayed if they match.

**Why you think this should be added:**

So that these redundant (for me) descriptions don't clutter my screen and don't steal my attention.

**Comment by low-batt** [comment]
IINA is a graphical user interface for the [mpv player](https://mpv.io/).  The mpv player is intended to primarily be used from the command line. It has a small controller on top of the video for basic control. The mpv project provides a library that can be embedded in other applications for video/audio playback. Using that library, IINA provides a full GUI for mpv as well as some other features such as downloading subtitles.

What this means for the case at hand is that IINA relies upon mpv to process subtitles. That means we must investigate mpv capabilities in this area.

Checking the [mpv manual](https://mpv.io/manual/stable/) I found [sub-filter-sdh](https://mpv.io/manual/stable/#options-sub-filter-sdh):
> **--sub-filter-sdh=<yes|no>**
>
>  &nbsp; &nbsp; &nbsp; &nbsp;Applies filter removing subtitle additions for the deaf or hard-of-hearing (SDH).
>  &nbsp; &nbsp; &nbsp; &nbsp;This is intended for English, but may in part work for other languages too. 
>  &nbsp; &nbsp; &nbsp; &nbsp;The intention is that it can be always enabled so may not remove all parts
>  &nbsp; &nbsp; &nbsp; &nbsp;added. It removes speaker labels (like MAN:), upper case text in parentheses
>  &nbsp; &nbsp; &nbsp; &nbsp;and any text in brackets.
>
> &nbsp; &nbsp; &nbsp; &nbsp; Default: no.

And [sub-filter-sdh-harder](https://mpv.io/manual/stable/#options-sub-filter-sdh-harder):
> **--sub-filter-sdh-harder=<yes|no>**
>
>  &nbsp; &nbsp; &nbsp; &nbsp;Do harder SDH filtering (if enabled by --sub-filter-sdh).
>  &nbsp; &nbsp; &nbsp; &nbsp;Will also remove speaker labels and text within parentheses using both lower
>  &nbsp; &nbsp; &nbsp; &nbsp;and upper case letters.
>
>   &nbsp; &nbsp; &nbsp; &nbsp;Default: no.

Many of the features of the mpv player that IINA does not directly support can still be used. To try out this mpv feature follow these steps:
- Open `IINA/Preferences`
- Click on `Advanced`
- Slide the `Enable advanced settings` toggle button to be on (blue)
- In the `Additional mpv options` section click on `+`
- A new entry appears in the table
- Double click on `name` and replace it with `sub-filter-sdh`
- Double click on `value` and replace it with `yes`
- Click on `+` again
- A new entry appears in the table
- Double click on `name` and replace it with `sub-filter-sdh-harder`
- Double click on `value` and replace it with `yes`
- Confirm the panel looks like the screenshot below
- **Restart IINA** to active the new settings

![issue-3958](https://user-images.githubusercontent.com/86170219/193494822-2423a6c9-854a-45da-9453-ee1ff9ce6875.jpg)


In the test I ran I found if I did not use the more agressive `sub-filter-sdh-harder` IINA did not remove the SDH subtitles in the sample SDH subtitles I found to test with.

Give those settings a try and report how well they work for you. If they don't do the job there are other options you can try such as [sub-filter-regex](https://mpv.io/manual/stable/#options-sub-filter-regex-):
> **--sub-filter-regex-...=...**
>
> &nbsp; &nbsp; &nbsp; &nbsp;Set a list of regular expressions to match on text subtitles,
> &nbsp; &nbsp; &nbsp; &nbsp;and remove any lines that match (default: empty).

See the manual for full details. **Note** the warning from mpv about that option being   _experimental_.

I'm expecting this issue should be treated by the IINA project as a request to add some sort of feature to make it easy to use the mpv feature to filter SDH subtitle entries.

**Comment by grishka** [comment]
Oh wow. Adding those two options did exactly what I wanted, thank you very much! I had no idea the code for such a seemingly niche feature was already written.

The issue is solved as far as I'm concerned. I'm leaving it open in case there are plans to add a GUI setting.

**Comment by low-batt** [comment]
Yes, please leave this issue open as a feature request to have IINA support this directly.

That probably will involve adding controls in the `Subtitle` section of `Preferences` as well as in the `SUBTITLES` tab of the `Quick Setting Panel`.

Glad this mpv feature proved to be useful.
